DataFormats/Detectors/TRD/include/DataFormatsTRD/TrapConfig3.h

```diff
@@ -0,0 +1,619 @@
+  uint32_t mBase;           //!< base of this registers block, i.e. TFL?? will have 0 and is in the range [0,kTrapRegistersSize]
+  uint32_t mWordNumber;     //!< word number offset, of size Nbits, in the block of registers
+  uint32_t mDataWordNumber; //!< offset, into "compressed" 32 bit words for the 32 bit word containing this register, offset from the base;
+  uint32_t mMask;           //!< mask to extract register from the word identified by WordNumber
+  uint32_t mMax;            //!< max is not the same as the mask, some values come in as 15 bit mask but are only 12 or 13 bits for max, IRQ values for example
+  uint32_t mShift;          //!< shift to extract the register
+  bool mIgnoreChange;       //!< we are not concerned with this value changing for purposes of the differential comparison.
+  ClassDefNV(TrapRegInfo, 1);
+};
+
+class TrapConfig3
+{
+  // class that is actually stored in ccdb.
+  // it holds a compressed version of what comes in the config events
+ public:
+  TrapConfig3();
+  ~TrapConfig3() = default;
+
+  // get a config register value by index, addr, and name, via mcm index
+  uint32_t getRegisterValue(uint32_t regidx, int mcmidx);
+  bool setRegisterValue(uint32_t data, uint32_t regidx, int mcmidx);
+  uint32_t getRegisterValueByIdx(uint32_t regidx, int mcmidx);
+  bool setRegisterValueByIdx(uint32_t data, uint32_t regidx, int mcmidx);
+  uint32_t getRegisterValueByAddr(uint32_t addr, int mcmidx);
+  bool setRegisterValueByAddr(uint32_t data, uint32_t addr, int mcmidx);
+  uint32_t getRegisterValueByName(const std::string& name, int mcmidx);
+  bool setRegisterValueByName(uint32_t data, const std::string& regname, int mcmidx);
+
+  // get a config register value by index, addr, and name, via sector/stack/layer/rob/mcm
+  // no setters for sector/stack/layer/rob/mcm as writing, we only write from retrieved config events.
+  uint32_t getRegisterValueByIdx(uint32_t regix, int sector, int stack, int layer, int rob, int mcm);
+  uint32_t getRegisterValueByAddr(uint32_t regaddr, int sector, int stack, int layer, int rob, int mcm);
+  uint32_t getRegisterValueByIdx(uint32_t regix, int detector, int rob, int mcm);
+  uint32_t getRegisterValueByAddr(uint32_t regaddr, int detector, int rob, int mcm);
+
+  // get a registers name by addres and index;
+  std::string getRegNameByAddr(uint16_t addr);
+  std::string getRegNameByIdx(unsigned int regidx);
+
+  // get a registers index (enum value) by address or name
+  int32_t getRegIndexByAddr(unsigned int addr);
+  int32_t getRegIndexByName(const std::string& name);
+
+  // get a registers address by index or name
+  int32_t getRegAddrByIdx(unsigned int regidx);
+  int32_t getRegAddrByName(const std::string& name);
+
+  // retrieve the trap register info for a particular register, no need to index into det/rob/mcm as these are constant over a config
+  o2::trd::TrapRegInfo* getTrapRegInfoByAddr(uint32_t addr);
+  o2::trd::TrapRegInfo* getTrapRegInfoByIdx(uint32_t idx);
```

Comment:
Feels like an antipattern to return a pointer to a class member because of semantics like 'who owns the memory afterward'

Reply:
It was there for debugging, done differently now, removed.